12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+      - MAY fail the request if its `amount` is much greater than the *base invoice amount*.
+      - MUST use the request's `amount` as the *base invoice amount*.
+  - otherwise:
+    - MUST fail the request if it does not contain `amount`.
+    - MUST use the request `amount` as the *base invoice amount*.
+  - if the offer has a `replace_invoice`:
+    - if the `payment_hash` refers to an unpaid invoice for the same `offer_id` and `payer_key`:
+      - MUST immediately expire/remove that unpaid invoice such that it cannot be paid in future.
+    - otherwise:
+      - MUST fail the request.
+
+## Rationale
+
+`payer_info` might typically contain information about the derivation of the
+`payer_key`.  This should not leak any information (such as using a simple
+BIP-32 derivation path); a valid system might be for a node to maintain a base
+payer key, and encode a 128-bit tweak here.  The payer_key would be derived by
+tweaking the base key with SHA256(payer_base_pubkey || tweak).
+
+`payer_note` allows you to compliment, taunt, or otherwise engrave
+graffiti into the invoice for all to see.
+
+Users can give a tip (or obscure the amount sent) by specifying an
+`amount` in their invoice request, even though the offer specifies an
+`amount`.  Obviously this will only be accepted by the recipient if
+the invoice request amount exceeds the amount it's expecting (i.e. its
+`amount` after any currency conversion, multiplied by `quantity` if
+any).  Note that for recurring invoices with `proportional_amount`
+set, the `amount` in the invoice request will be scaled by the time in
+the period; the sender should not attempt to scale it.
+
+`replace_invoice` allows the mutually-agreed removal of and old unpaid
+invoice; this can be used in the case of stuck payments.  If
+successful in replacing the stuck invoice, the sender may make a
+second payment such that it can prove double-payment should the
+receiver still accept the first, delayed payment.
+
+# Invoices
+
+Invoices are a request for payment, and when the payment is made 
+it can be combined with the invoice to form a cryptographic receipt.
+
+The human-readable prefix for invoices is `lni`.  It can be sent in
+response to an `invoice_request` or an `offer` with `send_invoice`
+using `onion_message` `invoice` field.
+
+1. `tlv_stream`: `invoice`
+2. types:
+    1. type: 3 (`chain`)
+    2. data:
+        * [`chain_hash`:`chain`]
+    1. type: 4 (`offer_id`)
+    2. data:
+        * [`sha256`:`offer_id`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 10 (`description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 16 (`paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 18 (`blindedpay`)
+    2. data:
+        * [`...*blinded_payinfo`:`payinfo`]
+    1. type: 19 (`blinded_capacities`)
+    2. data:
+        * [`...*u64`:`incoming_msat`]
```

Comment:
They're deliberately deconstructed: we use `paths` in offers, without `payinfo` (since it's just for sending a message).
`capacities` is an optional hint: I'm not sure if we're actually going to want that in practice.

Reply:
Right, that makes sense. One thing that is unclear to me though. Should we have one `payinfo` per blinded path or one per hop in each blinded path?

I believe we only need one per blinded path, and all hops in the blinded path should use the same `payinfo` to protect the anonymity of the channels inside the path. That also makes it lighter, especially when adding dummy hops at the end of a blinded path.

Reply:
Fixed!

Reply:
Actually I think the payer shouldn't need to know these channel infos. I suggest to put them inside the blinded route: https://github.com/lightning/bolts/pull/967

Reply:
You need to do both!  The payer needs to know the fees, but *also* we should eventually put something in the onion which says "I told them these are your fees!" to avoid probing.

Reply:
No, why would the payer need to know the fees if they are already inside the blinded route? The payer would just need to send the full amount to the first node of the blinded route and set dummy fees of zero for the rest of the blinded route and from there all the fees are read from the blinded route itself.
It also solves the problem of the recipient adding routing fees the payer has no control on, now the recipient pays these fees. I've always thought it was a problem (for bolt11 too with routing hints) that the payer if forced to pay fees for a part of the route they can't choose.

Reply:
OK, I see we went this way with blinded routes, which I'm reviewing now.  Let's change this to a simple whole-path fee.

Reply:
Actually, it is one per *path* already, which is correct.  I'll clarify the wording.

Reply:
If you're sending `x`, you need to send `5 + (1 + 1000e-6) * (5 + (1 + 1000e-6) * x) = (5 + (1 + 1000e-6) * 5) + (1 + (2001e-6)) * x`. So the aggregate fee is `5 + (1 + 1000e-6) * 5` base and `2001ppm`. The only imprecision is that at each hop the base and the proportional fees need to be rounded up but that should only amount to a total of just a few base msat and at most one extra ppm per hop.

Reply:
OK, I buy this.  I'll switch to a single amount, FTW.